waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -50,52 +51,71 @@ validateExactlyOneAppExists spec =
 validateDbIsPostgresIfPgBossUsed spec =
-  if isPgBossJobExecutorUsed spec && not (isPostgresUsed spec)
-    then
-      [ GenericValidationError
-          "Expected app.db.system to be PostgreSQL since there are jobs with executor set to PgBoss."
-      ]
-    else []
+  [ GenericValidationError
+      "Expected app.db.system to be PostgreSQL since there are jobs with executor set to PgBoss."
+    | isPgBossJobExecutorUsed spec && not (isPostgresUsed spec)
+  ]
 
 validateAuthUserEntityHasCorrectFieldsIfUsernameAndPasswordAuthIsUsed :: AppSpec -> [ValidationError]
 validateAuthUserEntityHasCorrectFieldsIfUsernameAndPasswordAuthIsUsed spec = case App.auth (snd $ getApp spec) of
   Nothing -> []
   Just auth ->
-    if Auth.UsernameAndPassword `notElem` Auth.methods auth
+    if not $ Auth.isUsernameAndPasswordAuthEnabled auth
       then []
       else
         let userEntity = snd $ AS.resolveRef spec (Auth.userEntity auth)
             userEntityFields = Entity.getFields userEntity
-            maybeUsernameField = find ((== "username") . Entity.Field.fieldName) userEntityFields
-            maybePasswordField = find ((== "password") . Entity.Field.fieldName) userEntityFields
-         in concat
-              [ case maybeUsernameField of
-                  Just usernameField
-                    | Entity.Field.fieldType usernameField == Entity.Field.FieldTypeScalar Entity.Field.String -> []
-                  _ ->
-                    [ GenericValidationError
-                        "Expected an Entity referenced by app.auth.userEntity to have field 'username' of type 'string'."
-                    ],
-                case maybePasswordField of
-                  Just passwordField
-                    | Entity.Field.fieldType passwordField == Entity.Field.FieldTypeScalar Entity.Field.String -> []
-                  _ ->
-                    [ GenericValidationError
-                        "Expected an Entity referenced by app.auth.userEntity to have field 'password' of type 'string'."
-                    ]
+         in concatMap
+              (validateField "app.auth.userEntity" userEntityFields)
+              [ ("username", Entity.Field.FieldTypeScalar Entity.Field.String, "String"),
+                ("password", Entity.Field.FieldTypeScalar Entity.Field.String, "String")
               ]
 
+validateExternalAuthAssociationEntityHasCorrectFieldsIfExternalAuthIsUsed :: AppSpec -> [ValidationError]
+validateExternalAuthAssociationEntityHasCorrectFieldsIfExternalAuthIsUsed spec = case App.auth (snd $ getApp spec) of
+  Nothing -> []
+  Just auth ->
+    if not $ Auth.isExternalAuthEnabled auth
+      then []
+      else case Auth.externalAuthAssociationEntity auth of
+        Nothing -> [GenericValidationError "Expected an Entity referenced by app.auth.externalAuthAssociationEntity to exist when using a social login method."]
+        Just externalAuthAssociationEntityRef ->
```

Comment:
Well this whole file is a product of us validating instead of parsing :D!
Have you not yet been with us @sodic when we debated how to do this validation of AppSpec? We had lengthy discussion + I explored multiple approaches in different PRs that I at the end abandoned, concluding that "correct" solutions were just too much work for us at the moment, and that "stupid" validation like this will work better for the time being.
If so, take a look at this PR here: https://github.com/wasp-lang/wasp/pull/455 .

Reply:
Yeah, this was before my time. Ok, works for me. Didn't know I was getting into a mine field with my comment :smile: